src/libraries/Microsoft.Extensions.Http/src/ActiveHandlerTrackingEntry.cs

```diff
@@ -85,9 +95,44 @@ private void Timer_Tick()
                     _timer.Dispose();
                     _timer = null;
 
-                    _callback(this);
+                    // Only invoke the callback if we successfully transition from 0 to Expired
+                    // This ensures we don't convert to expired if already disposed
+                    if (Interlocked.CompareExchange(ref _disposed, Expired, 0) == 0)
+                    {
+                        _callback(this);
+                    }
                 }
             }
         }
+
+        public void StopTimer()
+        {
+            lock (_lock)
+            {
+                _timer?.Dispose();
+                _timer = null;
+            }
+        }
+
+        public void Dispose()
+        {
+            // Try to transition from 0 to Disposed state
+            // If already in another state (Expired), do nothing further with handlers
+            int oldState = Interlocked.CompareExchange(ref _disposed, Disposed, 0);
+            if (oldState != 0)
+            {
+                // If the entry was already disposed or expired, exit
+                // If it was expired, the timer has already stopped and
+                // ExpiredHandlerTrackingEntry now owns both handler and scope
+                return;
+            }
+
+            StopTimer();
+
+            // When this entry is converted to an expired entry, we don't dispose the handlers
+            // to avoid a race with active clients. Let the ExpiredHandlerTrackingEntry handle handler disposal.
```

Comment:
At this moment, it's the other way around, we didn't convert to expired entry, and we should be the ones to dispose the handler here.